hyperopt/spark.py

```diff
@@ -0,0 +1,415 @@
+        elif is_success:
+            logger.debug("trial task {tid} succeeded, result is {r}"
+                         .format(tid=trial['tid'], r=data))
+            self._write_result_back(trial, result=data)
+        else:
+            logger.debug("trial task {tid} failed, exception is {e}"
+                         .format(tid=trial['tid'], e=str(data)))
+            self._write_exception_back(trial, e=data)
+
+    def launch_dispatcher(self):
+        def run_dispatcher():
+            start_time = timeit.default_timer()
+            while not self._fmin_done:
+                new_tasks = self._poll_new_tasks()
+
+                for trial in new_tasks:
+                    self._run_trial_async(trial)
+
+                elapsed_time = timeit.default_timer() - start_time
+
+                # In the future, timeout checking logic could be moved to `fmin`.
+                # For now, timeouts are specific to SparkTrials.
+                # When a timeout happens:
+                #  - Set `trials._fmin_cancelled` flag to be True.
+                #  - FMinIter checks this flag and exits if it is set to True.
+                if self.trials.timeout is not None and elapsed_time > self.trials.timeout and\
+                        not self.trials._fmin_cancelled:
+                    self.trials._fmin_cancelled = True
+                    self.trials._fmin_cancelled_reason = "fmin run timeout"
+                    self._cancel_running_trials()
+                    logger.warning("fmin cancelled because of " + self.trials._fmin_cancelled_reason)
+                time.sleep(1)
+
+            if self.trials._fmin_cancelled:
+                # Because cancelling fmin triggered, warn that the dispatcher won't launch
+                # more trial tasks.
+                logger.warning("fmin is cancelled, so new trials will not be launched.")
+
+            logger.debug("dispatcher thread exits normally.")
+
+        self._dispatcher_thread = threading.Thread(target=run_dispatcher)
+        self._dispatcher_thread.setDaemon(True)
+        self._dispatcher_thread.start()
+
+    @staticmethod
+    def _get_spec_from_trial(trial):
+        return base.spec_from_misc(trial['misc'])
+
+    @staticmethod
+    def _write_result_back(trial, result):
+        trial['state'] = base.JOB_STATE_DONE
+        trial['result'] = result
+        trial['refresh_time'] = coarse_utcnow()
+
+    @staticmethod
+    def _write_exception_back(trial, e):
+        trial['state'] = base.JOB_STATE_ERROR
+        trial['misc']['error'] = (str(type(e)), str(e))
+        trial['refresh_time'] = coarse_utcnow()
+
+    @staticmethod
+    def _write_cancellation_back(trial, e):
+        trial['state'] = base.JOB_STATE_CANCEL
+        trial['misc']['error'] = (str(type(e)), str(e))
+        trial['refresh_time'] = coarse_utcnow()
+
+    def _run_trial_async(self, trial):
+        def run_task_thread():
+            # TODO: use broadcast for `domain` object
```

Comment:
We actually were not able to do this because Spark's broadcast uses cPickle instead of cloudpickle.  @hanyucui would you mind removing this TODO?

I think this will be fine in general since the user's function and space should be fairly small objects.  The main worry would be if a user included a big dataset as an object in their function, rather than reading it inside their function.  I'll add a future TODO to the design doc list for checking the pickled object size.

Reply:
Sure, will remove it when I push a new patch.